### PR TITLE
Fixed #31246 -- Fixed locking models in QuerySet.select_for_update(of=()) for related fields and parent link fields with multi-table inheritance.

### PR DESCRIPTION
**Pull request by abhijeetviswa**
When a related model (with parents) is specified in `of` for select_for_update,
the `of` clause of the generated query references the field's parent model
instead of the field's model.

### CODE REVIEW

File: django/db/models/sql/compiler.py [hunk 1]
```diff
@@ -997,19 +997,21 @@ def _get_field_choices():
                     (path, klass_info)
                     for klass_info in klass_info.get('related_klass_infos', [])
                 )
+
+        def _get_col_from_model(klass_info):
```
[the remark this answers is not in the document]
Comment (abhijeetviswa) [reply django/db/models/sql/compiler.py]: Thanks. I'll fix these.

**Comment by felixxm** [comment]
@abhijeetviswa Thanks, please include tests and fix for multi-level inheritance, and add release notes for Django 2.2.10.

**Comment by abhijeetviswa** [comment]
@felixxm I'm not sure about the wording for the fix in the release notes. Could you please check it out?

Also, I'm not sure what more tests to add. I've modified `test_for_update_sql_model_inheritance_nested_ptr_generated_of` to test for related fields with inheritance and added `test_for_update_sql_multilevel_model_intehritance_ptr_generated_of` to test for multi-level inheritance.

File: tests/select_for_update/tests.py [hunk 2]
```diff
@@ -148,15 +150,30 @@ def test_for_update_sql_model_inheritance_ptr_generated_of(self):
     def test_for_update_sql_model_inheritance_nested_ptr_generated_of(self):
         with transaction.atomic(), CaptureQueriesContext(connection) as ctx:
             list(EUCity.objects.select_related('country').select_for_update(
-                of=('self', 'country__country_ptr',),
+                of=('self', 'country', 'country__country_ptr',),
```
Comment (felixxm) [inline tests/select_for_update/tests.py]: This is not related with using nested `_ptr`, so we should add a new test for this bugfix.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by carltongibson** [comment]
OK, looks good. 😅 

@felixxm Oracle failure seems unrelated — test didn't fail — but...

**Comment by felixxm** [comment]
Thanks for checking. Yes, Oracle failure is not related.
